banzai/bpm.py

```diff
@@ -8,7 +20,72 @@ def group_by_keywords(self):
         return None
 
     def do_stage(self, images):
+        images_to_remove = []
         for image in images:
-            bpm_slices = array_utils.array_indices_to_slices(image.bpm)
-            image.bpm[image.data[bpm_slices] >= float(image.header['SATURATE'])] = 2
+            try:
+                self.add_bpm_to_image(image)
+            except MissingBPMError:
```

Comment:
I don't think you need to catch an exception here. I think this stage should just load a dummy bpm and set L1BPMID (or whatever the keyword is supposed to be) to an empty string. Then the code here can just check if that header keyword is an empty string and remove the image if necessary. I think that should simplify the logic quite a bit and make it easier in the long run to remove that entirely.

Reply:
I've taken out the exception, but I've changed the logic a bit from what you proposed: first, an attempt is made to add a BPM. If `image.bpm` is still `None` after this (meaning that the BPM could not be loaded or was the wrong size), then the stage checks whether `pipeline_context.no_bpm` is `True`, and if so adds an empty BPM; otherwise, the image is removed.